### ISSUE TITLE
UWP IconImageSource from netstandard 2.0

### ISSUE THREAD

**Issue by jonkas2211** [xamarin/Xamarin.Forms#6534]
### Description
If you want to add your images in to the netstandard project and trie to add them in as a IconImageSource for a ToolbarItem then it does not appear on UWP. 
On Android it will appear as shown in the screenshot. 
If you add the image as an Image view inside your layout it will just work fine, so the ImageSource is not the problem.

### Steps to Reproduce

1. Add a image to the shared netstandard project.
2. Create an ImageSource of that image e.g. `var imageSource = ImageSource.FromResource("UWPToolbarItemsImageBug.Images.test.png", typeof(App).Assembly);`
3. Create a `ToolbarItem` and set the `IconImageSource` property
4. Add the ToolbarItem to the ToolbarItems

### Expected Behavior
Image should be shown in the toolbar.
### Actual Behavior
Image does not appear in the toolbar on UWP
### Basic Information

- Version with issue: lates (4.1.0.496342-pre2)
- Last known good version: -
- IDE: VS2019 & VS2017
- Platform Target Frameworks:
  - iOS:  not tested
  - Android: 9.0 
  - UWP:  16299, 17763
- Android Support Library Version: 28.0.0.1
- Nuget Packages: -
- Affected Devices: My Computer

### Screenshots

![grafik](https://user-images.githubusercontent.com/10948071/59514038-5e623100-8ebc-11e9-9eef-11de6c03f73d.png)

### Reproduction Link
[UWPToolbarItemsImageBug.zip](https://github.com/xamarin/Xamarin.Forms/files/3290908/UWPToolbarItemsImageBug.zip)

**Comment by jonkas2211** [comment]
Hallo @jfversluis,
can you give me a crudely estimation, when this will be fixed?

At the moment the alternative is, to copy all the images to the UWP project, which will produce redundancy of image files.

**Comment by jonkas2211** [comment]
After investigation, I found out the error can be caused, because the `Xamarin.Forms.Platform.UWP.StreamImageSourceHandler` is not an `IIconElementHandler`.

**Comment by jfversluis** [comment]
Hi @jonkas2211 unfortunately I do not have any estimation for you, sorry about that.

Since you have been investigating we are of course more than happy to see a pull request that fixes this issue. Have you actually verified that this is the problem?

**Comment by jonkas2211** [comment]
Hello @jfversluis,

The Method  `Xamarin.Forms.Platform.UWP.ImageExtensions.ToWindowsIconElementAsync(..)`
```c#
public static async Task<IconElement> ToWindowsIconElementAsync(this ImageSource source, CancellationToken cancellationToken = default(CancellationToken))
{
	if (source == null || source.IsEmpty)
		return null;

	var handler = Registrar.Registered.GetHandlerForObject<IIconElementHandler>(source);
	if (handler == null)
		return null;

	try
	{
		return await handler.LoadIconElementAsync(source, cancellationToken);
	}
	catch (OperationCanceledException)
	{
		// no-op
	}

	return null;
}
```

Is looking for the handler to implement `IIconElementHandler`. The `StreamImageSourceHandler` does not implement this, but the `FileImageSourceHandler` for example does.

I would love to create a pull request, but unfortunately I am no UWP expert and I dont know how to conver a Stream to a IconElement. I surely could save the Stream as a kind of image to the disk and relead it from an uri, but this is not how it supposed be?

**Comment by jonkas2211** [comment]
But I dont get any ToolbarItem to show up with other ImageSources... Is it completely broke on UWP?

**Comment by jonkas2211** [comment]
It might be a limitation of UWP

**Comment by jfversluis** [comment]
From what I can see this is simply not implemented in UWP indeed. The original implementation is from @mattleibow maybe he can shine his light on this?

I have been looking into it a little bit and the only possibility I see right now is to save the stream to a (temp) location and then load the images from there, but that doesn’t seem ideal as well. So then yes, it is a limitation on UWP.

**Comment by mattleibow** [comment]
Just spitballing here, but we might be able to do something with a bitmap that is a brush and use the brush. I see this was also added in a later version of UWP: https://docs.microsoft.com/en-us/uwp/api/windows.ui.xaml.controls.iconsourceelement

We might be able to do some things if we get a little bit creative.

**Comment by jonkas2211** [comment]
That seems to be a plan, but how would we handle targeting Windows 10, version 1809 (introduced v10.0.17763.0)?

**Comment by mattleibow** [comment]
It might be possible to use multitargeting to generate separate dlls so we can still support a desired target framework. But, the thing to do would be to bump the target version to the highest, but keep the minimum at the lowest. Then, use the `ApiInformation` type to see if the type or members are available (https://docs.microsoft.com/en-us/uwp/api/Windows.Foundation.Metadata.ApiInformation). This way the app can light up on newer platforms, and not crash on older ones.

**Comment by jonkas2211** [comment]
For reference look at this Issue on the Microsoft microsoft-ui-xaml repository: https://github.com/microsoft/microsoft-ui-xaml/issues/601

**Comment by jonkas2211** [comment]
So I tried to create a brush an set the Foreground from the IconElement, but it seems to end in a silent exception:
```csharp
public async Task<IconElement> LoadIconElementAsync(ImageSource imagesource, CancellationToken cancellationToken = new CancellationToken())
{
    IconElement image = null;
    if (imagesource is StreamImageSource streamSource)
    {
        image = new BitmapIcon();
        var brush = new ImageBrush();
        using (Stream stream = await((IStreamImageSource)streamSource).GetStreamAsync(cancellationToken))
        {
            if (stream == null)
                return null;

            var source = new BitmapImage();
            await source.SetSourceAsync(stream.AsRandomAccessStream());
            brush.ImageSource = source;
            image.SetForeground(brush); // throws silent Argument Exception seen in output
        }
    }
    return image;
}
```

**Comment by mattleibow** [comment]
If you wrap the code in a try catch, is it possible to see what it is? Maybe the async is causing the silence.

**Comment by jonkas2211** [comment]
Thanks @mattleibow if things could always be easy as that.
So the exception is saying, that the ImageBrush is not a Foreground Type: `type is not a Foregroundable type`

**Comment by jonkas2211** [comment]
Hallo @samhouts,
this issue seems to be fixed with pull request #8147.

**Comment by jfversluis** [comment]
Thanks for letting us know! Closing this one then
